### ISSUE TITLE
no-useless-assertion: false positive when disabling no-misused-generics

### ISSUE THREAD

**Issue by AlCalzone** [fimbullinter/wotan#625]
Consider the following code:
```ts
/* wotan-disable no-misused-generics */
function foo<T = unknown>(): T {
	return ({} as any) as T;
}

const bar = foo() as string;
console.log(bar.length);
```
`wotan` complains about the `as string` being a useless assertion. However when I remove it, the type of `bar` changes to `unknown`, making the last line illegal.

**Comment by ajafff** [comment]
I remember seeing this in the TSLint repo. This was discussed in https://github.com/palantir/tslint/issues/3540 and https://github.com/microsoft/TypeScript/issues/20681

TL;DR: this not only occurs when disabling `no-misused-generics`, but every time a call(like) expression has no inference candidate for a type parameter used in the return type. Instead of using the declared fallback the type is inferred from the contextual type. In this case the contextual type comes from the type assertion. If you remove the assertion, there's no contextual type and therefore the type parameter default applies.

You can work around this design limitation by providing the contextual type by any other means:

```ts
const bar: string = foo(); // 'T' is inferred as 'string'
```

---

I don't think this can be fixed in this project. If you care enough about this issue, please consider opening a new issue with TypeScript.

**Comment by AlCalzone** [comment]
> If you care enough about this issue

I don't, since I can just pass the intended type in the angled brackets. I just thought the behavior is weird.
